### PR TITLE
Dynamically allocate gcs mavlink backends

### PR DESCRIPTION
**Pull request by peterbarker**
@WickedShell expressed interest in this.  So I spruced it up and did the Copter stuff.

This PR causes us to allocate GCS_MAVLink backends only as required.  In master the backends are held in an array in the vehicle GCS objects:

```
    GCS_MAVLINK_Plane _chan[MAVLINK_COMM_NUM_BUFFERS];
```

https://github.com/ardupilot/ardupilot/blob/master/ArduPlane/GCS_Plane.h#L31

This PR changes things so that that is instead an array of pointers, and we `new` backends based on the `SERIALx_PROTOCOL` parameters.  The streamrate protocol parameters which were formerly held in these object are now held in a different class.

Doing things this was should save us some memory.  Unexpectedly it also current seems to save us some flash:

```
wip/dynamically-allocate-gcs_mavlink:
  bin/arduplane  983624  1152  195640  1180416
  3 mavlink backends:
    MEMINFO {brkval : 0, freemem : 65535, freemem32 : 89864}
  1 mavlink backend:
    MEMINFO {brkval : 0, freemem : 65535, freemem32 : 92600}

master:
  bin/arduplane  984208  1152  195640  1181000
  3 mavlink backends configured using SERIALx_PROTOCOL:
    MEMINFO {brkval : 0, freemem : 65535, freemem32 : 89032}
  1 mavlink backend configured:
    MEMINFO {brkval : 0, freemem : 65535, freemem32 : 90832}
```

Drawback: null-pointer checking
Drawback: run-time null-pointer checking

I'm after feedback on this concept.

I will highlight one piece I'm not happy about in a code review.

### CODE REVIEW

File: ArduCopter/GCS_Copter.cpp [hunk 1]
```diff
@@ -183,3 +183,24 @@ void GCS_Copter::update_vehicle_sensor_status_flags(void)
     }
 #endif
 }
+
+bool GCS_Copter::create_gcs_mavlink_backend(AP_SerialManager &serial_manager,
```
Comment (peterbarker) [inline ArduCopter/GCS_Copter.cpp]: I'm unhappy at cloning this method for each vehicle.

I *think* I could probably store the pointers in the base class - but I'd then have to cast them back into their true (subclass) pointer type in the `chan(uint8_t)` method.  If I were to do that this method would just return the `new`'d object.

Does that seem reasonable to anybody?

File: ArduCopter/compassmot.cpp [hunk 2]
```diff
@@ -33,7 +33,12 @@ MAV_RESULT Copter::mavlink_compassmot(mavlink_channel_t chan)
         ap.compass_mot = true;
     }
 
-    GCS_MAVLINK_Copter &gcs_chan = gcs().chan(chan-MAVLINK_COMM_0);
```
Comment (peterbarker) [inline ArduCopter/compassmot.cpp]: Would be nice if this method actually took a `GCS_MAVLink*`

File: libraries/GCS_MAVLink/GCS.cpp [hunk 3]
```diff
@@ -48,18 +48,15 @@ void GCS::send_to_active_channels(uint32_t msgid, const char *pkt)
         return;
     }
     for (uint8_t i=0; i<num_gcs(); i++) {
-        GCS_MAVLINK &c = chan(i);
-        if (!c.initialised) {
+        GCS_MAVLINK *c = chan(i);
```
Comment (WickedShell) [inline libraries/GCS_MAVLink/GCS.cpp]: Not critical, but the diff could have been minimized in this area.
```suggestion
        GCS_MAVLINK &c = *chan(i);
```

Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS.cpp]: Done, thanks.

File: ArduCopter/GCS_Copter.h [hunk 4]
```diff
@@ -10,13 +10,20 @@ class GCS_Copter : public GCS
 public:
 
     // return the number of valid GCS objects
-    uint8_t num_gcs() const override { return ARRAY_SIZE(_chan); };
+    uint8_t num_gcs() const override { return _num_gcs; };
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply ArduCopter/GCS_Copter.h]: Yes!  Done.

File: ArduCopter/GCS_Copter.cpp (hunk 1 above)
Comment (peterbarker) [reply ArduCopter/GCS_Copter.cpp]: I've factored this and moved it up, and cast the pointer on the getter.

File: ArduCopter/compassmot.cpp (hunk 2 above)
Comment (peterbarker) [reply ArduCopter/compassmot.cpp]: Done in a separate PR.

**Comment by peterbarker** [comment]
```
pr/dynamically-allocate-gcs_mavlink
  bin/arduplane  984428  1152  195644  1181224
  1 mavlink backend:
   freemem32 : 92600
  2 mavlink backend:
   freemem32 : 92144
  3 mavlink backends:
   freemem32 : 89888


master
  bin/arduplane  984348  1152  195644  1181144
  1 mavlink backend:
   freemem32 : 90816
  2 mavlink backend:
   freemem32 : 90816
  3 mavlink backends:
   freemem32 : 89016
```

**Comment by peterbarker** [comment]
No longer saving flash.  But still saving memory.

File: AntennaTracker/GCS_Tracker.cpp [hunk 5]
```diff
@@ -4,7 +4,6 @@
 void GCS_Tracker::request_datastream_position(const uint8_t sysid, const uint8_t compid)
 {
     for (uint8_t i=0; i < num_gcs(); i++) {
-        if (gcs().chan(i).initialised) {
             // request position
```
Comment (WickedShell) [inline AntennaTracker/GCS_Tracker.cpp]: Whitespace. There is a lot of off whitespace now...

Comment (peterbarker) [reply AntennaTracker/GCS_Tracker.cpp]: This was quite intentional to make it easier to review.  It  can be fixed with an NFC follow-up.

**Comment by peterbarker** [comment]
>     * _GCS_MAVLink: routing example: use GCS_Dummy in place of custom GCS_: no such thing happens in this commit?

Ah, that change got pulled out of this PR and is already merged.  I've folded this patch into another.

>     * _AP_Airspeed: var_info is now in GCS_MAVLINK_Parameters_ and _AP_AHRS: var_info is now in GCS_MAVLINK_Parameters_ should have `examples: ` added to it (as the others following do)

Fixed.

File: libraries/GCS_MAVLink/GCS.h [hunk 6]
```diff
@@ -282,7 +282,16 @@ class MissionItemProtocol_Rally : public MissionItemProtocol {
                                 mavlink_mission_item_int_t &ret_packet) override WARN_IF_UNUSED;
 
     static MAV_MISSION_RESULT convert_MISSION_ITEM_INT_to_RallyLocation(const mavlink_mission_item_int_t &cmd, RallyLocation &ret) WARN_IF_UNUSED;
+};
+
+#define NUM_STREAM_RATES 10
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS.h]: Done

File: libraries/GCS_MAVLink/GCS.h [hunk 7]
```diff
@@ -282,7 +282,16 @@ class MissionItemProtocol_Rally : public MissionItemProtocol {
...
+class GCS_MAVLINK_Parameters
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS.h]: Yes, I agree.  I actually attempted that - it's harder than it looks due to circular dependencies IIRC.

I definitely don't want to do that sort of thing on this PR.

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 8]
```diff
@@ -73,56 +73,30 @@ uint8_t GCS_MAVLINK::mavlink_private = 0;
...
-    if (!valid_channel(mav_chan)) {
-        return;
-    }
-
-    _port = port;
-    chan = mav_chan;
-
-    mavlink_comm_port[chan] = _port;
-    _queued_parameter = nullptr;
+    // AP_Param::setup_object_defaults(this, var_info);
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Removed.

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 9]
```diff
@@ -2266,17 +2242,48 @@ void GCS::update_receive(void)
...
+    AP_HAL::UARTDriver *uart = serialmanager.find_serial(AP_SerialManager::SerialProtocol_MAVLink, 0);
+    if (uart == nullptr) {
+        // this is probably not going to end well.
+        return;
+    }
+    create_gcs_mavlink_backend(uart);
+}
+
+
+bool GCS::create_gcs_mavlink_backend(AP_HAL::UARTDriver *uart)
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Made it void.

File: libraries/GCS_MAVLink/GCS_Dummy.h [hunk 10]
```diff
@@ -23,6 +23,11 @@ const struct GCS_MAVLINK::stream_entries GCS_MAVLINK::all_stream_entries[] {};
  */
 class GCS_MAVLINK_Dummy : public GCS_MAVLINK
 {
+public:
+        using GCS_MAVLINK::GCS_MAVLINK;
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Dummy.h]: Fixed..

File: libraries/GCS_MAVLink/GCS_Dummy.h [hunk 11]
```diff
@@ -53,10 +58,21 @@ extern const AP_HAL::HAL& hal;
...
+
+protected:
+
+    GCS_MAVLINK_Dummy *new_gcs_mavlink_backend(GCS_MAVLINK_Parameters &params,
+                                               AP_HAL::UARTDriver *uart) override {
+        return new GCS_MAVLINK_Dummy(params, uart);
+    }
+
+private:
+    GCS_MAVLINK_Dummy *dummy_backend;
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Dummy.h]: It's not.  This didn't keep up with other changes.  Changed it to cast it as the other subclasses do.

**Comment by peterbarker** [comment]
This one can't go in until we sort out our stoppages on master; do disruptive at this point.

There's an additional problem in this PR where we're doing the wrong range check on the sanity check.  That's fixed in a different PR which I'll rebase this on top of once it's merged.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (OXINARF) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: This doesn't actually answer my question. How are the parameters being setup if we aren't calling the setup method anymore?

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 12]
```diff
@@ -133,29 +107,35 @@ GCS_MAVLINK::setup_uart(const AP_SerialManager& serial_manager, AP_SerialManager
...
+
+    snprintf(_perf_update_name, sizeof(_perf_update_name), "GCS_Update_%u", chan);
+    _perf_update = hal.util->perf_alloc(AP_HAL::Util::PC_ELAPSED, _perf_update_name);
 
-    AP_SerialManager::SerialProtocol mavlink_protocol = serialmanager_p->get_mavlink_protocol(mav_chan);
+    AP_SerialManager::SerialProtocol mavlink_protocol = serialmanager_p->get_mavlink_protocol(chan);
     mavlink_status_t *status = mavlink_get_channel_status(chan);
     if (status == nullptr) {
-        return;
+        return true; // ??!?!?!
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: The entire library seems to be built around that point being able to come back as nullptr - I decided to roll with that.

[the remark this answers is not in the document]
Comment (OXINARF) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: If status is null then something is wrong and we lose important settings - like the signing key (which puts the vehicle in risk)!

I think this should return false here, but if true is the correct answer then the comment should be removed...

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: rebasing has not been kind to this PR.

Added relevant call.

Thanks for reviewing this.  Sadly this won't get merged for a while - until we sort out what's happening with the Solos, probably.

**Comment by peterbarker** [comment]
On Tue, 25 Jun 2019, Francisco Ferreira wrote:

> I think this should return false here, but if true is the correct answer then the comment should be removed...

You might gather I was rather dubious about returning true there myself.

As things currently stand, behaviour is preserved, 'though?

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (OXINARF) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: > rebasing has not been kind to this PR.

I'm guessing rerere isn't helping 😛 

> Sadly this won't get merged for a while - until we sort out what's happening with the Solos, probably.

How does this specifically impact Solos?

**Comment by peterbarker** [comment]
@OXINARF All addressed now?

I wouldn't mind changing the null-ptr-check, but I'm not sure it's a shoe-in.  If we're going to absolutely require it exists it probably makes sense to keep a reference to it in `*this`, and use that reference everywhere instead of asking for it every time we want it and doing a nullptr check.

**Comment by OXINARF** [comment]
> You might gather I was rather dubious about returning true there myself.
>
> As things currently stand, behaviour is preserved, 'though?

Yes, doesn't mean current behavior is good though. It should likely be clarified in a dev call.

> I wouldn't mind changing the null-ptr-check, but I'm not sure it's a shoe-in. If we're going to absolutely require it exists it probably makes sense to keep a reference to it in *this, and use that reference everywhere instead of asking for it every time we want it and doing a nullptr check.

The point against doing that is that you would be "interfering" with the MAVLink library implementation, nothing says that the pointer has to always be the same.

But it should always return a status - if it doesn't, something is very wrong (by the way, MAVLink library will attempt to dereference a nullptr in that case).

> All addressed now?

Other than the above, yes.

**Comment by peterbarker** [comment]
On Tue, 25 Jun 2019, Francisco Ferreira wrote:

> How does this specifically impact Solos?

It does not, but tridge doesn't want churn in master until we work out 
what's going on with Solos.

**Comment by OXINARF** [comment]
@peterbarker can we merge this now?

File: APMrover2/GCS_Rover.h [hunk 13]
```diff
@@ -9,24 +9,21 @@ class GCS_Rover : public GCS
...
-    uint8_t num_gcs() const override { return ARRAY_SIZE(_chan); };
-
     // return GCS link at offset ofs
-    GCS_MAVLINK_Rover &chan(uint8_t ofs) override {
-        if (ofs >= num_gcs()) {
+    GCS_MAVLINK_Rover *chan(const uint8_t ofs) override {
+        if (ofs > _num_gcs) {
             AP::internalerror().error(AP_InternalError::error_t::gcs_offset);
-            ofs = 0;
+            return nullptr;
```
Comment (peterbarker) [inline APMrover2/GCS_Rover.h]: @OXINARF Should we change this to return chan[0] as we were?  We don't need to change this behaviour as part of this PR.

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 14]
```diff
@@ -1883,17 +1861,47 @@ void GCS::update_receive(void)
...
+        // this is probably not going to end well.
+        return;
     }
+    create_gcs_mavlink_backend(uart);
+}
+
+
+void GCS::create_gcs_mavlink_backend(AP_HAL::UARTDriver *uart)
+{
+    _chan[_num_gcs] = new_gcs_mavlink_backend(chan_parameters[_num_gcs], uart);
```
Comment (OXINARF) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: We need to check that `_num_gcs` is less than the array size.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (OXINARF) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: This is still unresolved. By the way, as written in https://github.com/ArduPilot/ardupilot/pull/11617#issuecomment-505384345, we'll end up with a nullptr dereference if we return true here.

File: APMrover2/GCS_Rover.h (hunk 13 above)
Comment (OXINARF) [reply APMrover2/GCS_Rover.h]: I don't think we should, channel 0 might not exist either. 

I think we just need to fix one case where we assume that at least channel 0 exists (https://github.com/ArduPilot/ardupilot/pull/11617/files/ff81a5248f1e4c4ec73f5b19d2ed660d16aa1a06#diff-184f24112b8a738e36349802bd5628eeR345) and then we are safe.

File: libraries/GCS_MAVLink/GCS.h [hunk 15]
```diff
@@ -998,16 +1013,25 @@ class GCS
...
+    virtual GCS_MAVLINK *new_gcs_mavlink_backend(GCS_MAVLINK_Parameters &params,
+                                                 AP_HAL::UARTDriver *uart) = 0;
+
     uint32_t control_sensors_present;
     uint32_t control_sensors_enabled;
     uint32_t control_sensors_health;
     void update_sensor_status_flags();
     virtual void update_vehicle_sensor_status_flags() {}
 
+    GCS_MAVLINK_Parameters chan_parameters[MAVLINK_COMM_NUM_BUFFERS];
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS.h]: Later :-)

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Changed to return false.

File: APMrover2/GCS_Rover.h (hunk 13 above)
Comment (peterbarker) [reply APMrover2/GCS_Rover.h]: Fixed.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 14 above)
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Fixed

**Comment by peterbarker** [comment]
I've tested that the serial rate parameters do work the same as you move forward and backward between master and this code.

**Pull request merged**